Added support for calling GAScriptObject as a function

b123400 opened the pull request:
"callFunction" only allows a function to be called if it is a property of an object. "callAsFunction" allows functions to be called directly, which is more flexible.

One good application would be using callback to handle the no-return-value problem.

``` Objective-C
-(void)getValueWithCallback:(GAScriptObject*)callback{
    [callback callAsFunctionWithArguments:@[@"Hello"]];
}
```

``` Javascript
GAJavaScript.performSelector( 'getValueWithCallback:', function(value){
    // value == "Hello"
});
```

b123400 committed:
Added support for function as a GAScriptObject
Able to call GAScriptObject as a function

b123400 committed:
Script object will not cause crashes when it is not released main thread.
Change undefined-->nil to NSNull, so it can be put into array

b123400 committed:
stringForJavaScript already making an array, so dont need to wrap it.

b123400 committed:
Added callAsFunction to README as a solution to the no-return-value problem

b123400 committed:
Merge remote-tracking branch 'upstream/master'

newyankeecodeshop commented:
This is a good idea. Since functions are objects though, you could invoke the function by using the "call" or "apply" method?
    [GAScriptObject callFunction:@"apply" withObject:arrayOfArgs]

b123400 committed:
Expose js runtime path and added custom injection

This will allow application to manage when to inject the script. The
current method cannot make sure js is injected before your code runs

b123400 committed:
Set the delegate manually to avoid crash with BlocksKit

b123400 committed:
Handle the case where the engine (webview) dealloc before script objects
